types.cc

```diff
@@ -624,7 +624,13 @@ void write_collection_size(bytes::iterator& out, int size, cql_serialization_for
 }
 
 bytes_view read_collection_value(bytes_view& in, cql_serialization_format sf) {
-    auto size = sf.using_32_bits_for_collections() ? read_simple<int32_t>(in) : read_simple<uint16_t>(in);
+    int32_t size = sf.using_32_bits_for_collections() ? read_simple<int32_t>(in) : read_simple<uint16_t>(in);
+    if (size == -2) {
+        throw exceptions::invalid_request_exception("unset value is not supported inside collections");
+    }
+    if (size < 0) {
```

Comment:
```
[bytes]        A [int] n, followed by n bytes if n >= 0. If n < 0,
               no byte should follow and the value represented is `null`.

[value]        A [int] n, followed by n bytes if n >= 0.
               If n == -1 no byte should follow and the value represented is `null`.
               If n == -2 no byte should follow and the value represented is
               `not set` not resulting in any change to the existing value.
               n < -2 is an invalid value and results in an 

6.11 list
  A [int] n indicating the number of elements in the list, followed by n
  elements.  Each element is [bytes] representing the serialized value.
```

Technically a list consists of many [bytes], so everything < 0 is a null no matter whether it is -1, -2 or -1000.
I could check only for nulls but the user could make a mistake where they send send unset value and it would be confusing for them to receive message about a null, so I think it's good to also recognize this case even if it isn't technically correct.

Reply:
ok, makes sense, thanks for clarification